components/web_discovery/browser/web_discovery_service.cc

```diff
@@ -92,6 +99,61 @@ void WebDiscoveryService::OnConfigChange() {
   credential_manager_->JoinGroups();
 }
 
-void WebDiscoveryService::OnPatternsLoaded() {}
+void WebDiscoveryService::OnPatternsLoaded() {
+  if (!content_scraper_) {
+    content_scraper_ =
+        std::make_unique<ContentScraper>(server_config_loader_.get());
+  }
+}
+
+bool WebDiscoveryService::ShouldExtractFromPage(
+    const GURL& url,
+    content::RenderFrameHost* render_frame_host) {
+  if (!content_scraper_) {
+    return false;
+  }
+  const auto* matching_url_details =
+      server_config_loader_->GetLastPatterns().GetMatchingURLPattern(url,
+                                                                     false);
+  if (!matching_url_details) {
+    return false;
+  }
+  VLOG(1) << "URL matched pattern " << matching_url_details->id << ": " << url;
+  if (IsPrivateURLLikely(url, matching_url_details)) {
+    return false;
+  }
+  return true;
+}
+
+void WebDiscoveryService::StartExtractingFromPage(
+    const GURL& url,
+    mojo::Remote<mojom::DocumentExtractor> document_extractor) {
+  auto remote_id =
+      document_extractor_remotes_.Add(std::move(document_extractor));
+  // We use a WeakPtr within the ContentScraper for callbacks from the renderer,
+  // so using Unretained is fine here.
+  CHECK(content_scraper_);
+  content_scraper_->ScrapePage(
```

Comment:
basically content_scraper looks like a good candidate to be used via `base::SequenceBound`, both it's public functions look like they should be async

Reply:
yes, it could be, though I don't think `ScrapePage` warrants a task runner considering it already dispatches the compute-heavy operations to the renderer process. `ParseAndScrapePage` already dispatches the HTML parsing/extraction to a task runner.

In addition, the class uses the `ServerConfigLoader` to get scraping rules and other attributes, so the caller would be responsible for retrieving the necessary info ahead of time, making the class slightly less convenient to use.

The status quo allows us to choose whether a separate task is needed, depending on the chosen operation.

Reply:
but since it returns async anyway wouldn't it be simpler to make it sequence bound and then we don't have to worry about it? Seems like added complexity to maybe have a task runner internally? Not a blocker, but I don't really see the advantage of not making it sequence bound

Reply:
also just passing to the renderer is not sufficient because you also don't want to block the main thread there. Please make sure that any non-trivial operations are annotated with `base::AssertLongCPUWorkAllowed`